File: rig/bones.js
// Spline-bone deformation engine (Marduk rig). Single source of truth for the
// runtime; rig/tools/bones.py is the Python mirror. See rig/BONES.md.
//
// A "bone" deforms a 2D shape along a skeleton:
//   bone = { polys, rest, bent, xmn, xmx, ymid }
//     polys  : array of point-arrays [[x,y],...] — the shape, in the bone's local frame
//     rest   : straight skeleton polyline (the shape at parameter 0)
//     bent   : curled skeleton polyline, SAME point count (the shape at parameter 1)
//     xmn,xmx: the shape's extent ALONG the bone (maps to arc-length on the skeleton)
//     ymid   : the shape's centre ACROSS the bone (Y offsets are measured from here)
// A caller places the whole bone in the scene with its own translate+rotate.
const Bones = (function(){
  const TAU = Math.PI*2;
  const norm = a => (((a+Math.PI)%TAU)+TAU)%TAU - Math.PI;      // wrap angle to (-pi, pi]

  // Arc-length lookup over a polyline. at(s) -> {p:[x,y], t:[ux,uy]} (point + unit tangent).
  function arclen(pts){
    const seg=[], cum=[0]; let L=0;
    for(let i=0;i<pts.length-1;i++){
      const dx=pts[i+1][0]-pts[i][0], dy=pts[i+1][1]-pts[i][1], l=Math.hypot(dx,dy);
      seg.push([dx,dy,l]); L+=l; cum.push(L);
    }
    return { L, at(s){ s=Math.max(0,Math.min(L,s)); let i=0; while(i<seg.length-1 && cum[i+1]<s) i++;
      const f=(s-cum[i])/Math.max(seg[i][2],1e-9);
      return { p:[pts[i][0]+seg[i][0]*f, pts[i][1]+seg[i][1]*f],
               t:[seg[i][0]/Math.max(seg[i][2],1e-9), seg[i][1]/Math.max(seg[i][2],1e-9)] };
    }};
  }

  // Interpolate the skeleton rest->bent at t by accumulating TURNING angles (relative),
  // so the chain rolls up progressively. Position-lerp would collapse through the straight
  // chord; absolute-angle lerp spikes when the skeleton hooks back on itself.
  function interpSkel(rest, bent, t){
    const aR=[],lR=[],aB=[],lB=[];
    for(let i=0;i<rest.length-1;i++){
      aR.push(Math.atan2(rest[i+1][1]-rest[i][1], rest[i+1][0]-rest[i][0]));
      lR.push(Math.hypot(rest[i+1][0]-rest[i][0], rest[i+1][1]-rest[i][1]));
      aB.push(Math.atan2(bent[i+1][1]-bent[i][1], bent[i+1][0]-bent[i][0]));
      lB.push(Math.hypot(bent[i+1][0]-bent[i][0], bent[i+1][1]-bent[i][1]));
    }
    let h = aR[0] + t*norm(aB[0]-aR[0]);                         // first-segment heading
    const V=[[ rest[0][0]+(bent[0][0]-rest[0][0])*t, rest[0][1]+(bent[0][1]-rest[0][1])*t ]];
    for(let i=0;i<aB.length;i++){
      if(i>0) h += t*norm(aB[i]-aB[i-1]);                        // a fraction of each joint's final turn
      const l=(1-t)*lR[i]+t*lB[i], p=V[V.length-1];
      V.push([ p[0]+l*Math.cos(h), p[1]+l*Math.sin(h) ]);
    }
    return V;
  }

  // Inkscape "Bend" deformation: each point's X (along) -> arc-length on the skeleton,
  // its Y (across, from ymid) -> perpendicular offset. Returns an SVG path 'd'.
  function bendAlong(polys, skel, xmn, xmx, ymid){
    const A=arclen(skel), Xr=xmx-xmn, out=[];
    for(const poly of polys){ let d='M ';
      for(let j=0;j<poly.length;j++){
        const x=poly[j][0], y=poly[j][1], r=A.at((x-xmn)/Xr*A.L);
        d += (j?'L ':'') + (r.p[0]+(y-ymid)*(-r.t[1])).toFixed(2)+','+(r.p[1]+(y-ymid)*(r.t[0])).toFixed(2)+' ';
      }
      out.push(d+'Z');
    }
    return out.join(' ');
  }

  // deform a bone at t (0..1) -> path 'd';  skelPath -> the live driving skeleton (overlay/debug).
  const deform   = (bone, t) => bendAlong(bone.polys, interpSkel(bone.rest, bone.bent, t), bone.xmn, bone.xmx, bone.ymid);
  const skelPath = (rest, bent, t) => 'M ' + interpSkel(rest,bent,t).map(p=>p[0].toFixed(2)+','+p[1].toFixed(2)).join(' L ');

  // Frame at the skeleton's TIP (last point) at t, in the bone's LOCAL frame:
  // { p:[x,y] tip position, ang: tangent angle of the last segment }. Used to
  // parent a child (e.g. the hand) to the bone's far end (e.g. the wrist).
  function tipFrame(bone, t){
    const V=interpSkel(bone.rest, bone.bent, t), n=V.length, a=V[n-1], b=V[n-2];
    return { p:[a[0],a[1]], ang:Math.atan2(a[1]-b[1], a[0]-b[0]) };
  }

  return { norm, arclen, interpSkel, bendAlong, deform, skelPath, tipFrame };
})();
if (typeof module!=='undefined') module.exports = Bones;

File: rig/tools/bones.py
#!/usr/bin/env python3
"""Spline-bone deformation engine (Python mirror of rig/bones.js). See rig/BONES.md.

A bone deforms a 2D shape along a skeleton:
  bone = {polys, rest, bent, xmn, xmx, ymid}   (rest/bent same point count)
Runtime uses turning-angle interpolation so the chain rolls up progressively."""
import numpy as np, math

def arclen(pts):
    """Arc-length lookup over a polyline. at(s) -> (point, unit_tangent)."""
    seg=np.diff(pts,axis=0); sl=np.hypot(seg[:,0],seg[:,1]); cum=np.r_[0,np.cumsum(sl)]; L=cum[-1]
    def at(s):
        s=np.clip(s,0,L); i=min(int(np.searchsorted(cum,s))-1,len(seg)-1); i=max(i,0)
        f=(s-cum[i])/max(sl[i],1e-9)
        return pts[i]+seg[i]*f, seg[i]/max(sl[i],1e-9)
    return at,L

def interp_skel(rest, bent, t):
    """Interpolate skeleton rest->bent by accumulating TURNING angles (relative)."""
    def dl(P): s=np.diff(P,axis=0); return np.arctan2(s[:,1],s[:,0]), np.hypot(s[:,0],s[:,1])
    aR,lR=dl(rest); aB,lB=dl(bent)
    turn=np.zeros(len(aB))
    for i in range(1,len(aB)): turn[i]=(aB[i]-aB[i-1]+math.pi)%(2*math.pi)-math.pi
    h=aR[0]+t*((aB[0]-aR[0]+math.pi)%(2*math.pi)-math.pi)
    V=[(1-t)*rest[0]+t*bent[0]]
    for i in range(len(aB)):
        if i>0: h+=t*turn[i]
        l=(1-t)*lR[i]+t*lB[i]; V.append(V[-1]+l*np.array([math.cos(h),math.sin(h)]))
    return np.array(V)

def bend_along(polys, skel, xmn, xmx, ymid):
    """Inkscape Bend: X (along) -> arc-length on skel, Y (from ymid) -> perpendicular. -> path 'd'."""
    at,L=arclen(skel); Xr=xmx-xmn; parts=[]
    for poly in polys:
        bp=[]
        for x,y in poly:
            P,tn=at((x-xmn)/Xr*L); bp.append((P[0]+(y-ymid)*(-tn[1]), P[1]+(y-ymid)*tn[0]))
        parts.append('M '+' L '.join(f'{x:.2f},{y:.2f}' for x,y in bp)+' Z')
    return ' '.join(parts)

def deform(bone, t):
    """Deform a bone {polys,rest,bent,xmn,xmx,ymid} (numpy arrays) at t -> path 'd'."""
    return bend_along(bone['polys'], interp_skel(bone['rest'], bone['bent'], t),
                      bone['xmn'], bone['xmx'], bone['ymid'])

def tip_frame(bone, t):
    """Frame at the skeleton tip (last point) at t, in the bone's LOCAL frame ->
    (tip_point, tangent_angle). Used to parent a child to the bone's far end."""
    V=interp_skel(bone['rest'], bone['bent'], t)
    return V[-1], math.atan2(V[-1][1]-V[-2][1], V[-1][0]-V[-2][0])
